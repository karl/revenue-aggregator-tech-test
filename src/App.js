import React, { useEffect, useState } from "react";
import "./App.css";

const createLogger = (name) => {
  const log = (level, message, extra) => {
    if (process.env.NODE_ENV === "test") {
      return;
    }

    if (extra !== undefined) {
      console[level](`[${name}] ${message}`, extra);
    } else {
      console[level](`[${name}] ${message}`);
    }
  };

  return {
    info: (message, extra) => {
      log("info", message, extra);
    },
    warning: (message, extra) => {
      log("warn", message, extra);
    },
    error: (message, extra) => {
      log("error", message, extra);
    },
  };
};

const logger = createLogger("App");

const formatNumber = (number) =>
  new Intl.NumberFormat("en", { minimumFractionDigits: 2 }).format(number);

const App = () => {
  const [state, setState] = useState({
    uiState: "LOADING",
    sortedProducts: undefined,
  });
  const [filter, setFilter] = useState("");

  useEffect(() => {
    const getData = async () => {
      try {
        logger.info("Getting data");
        const results = await Promise.all([
          fetch("./api/branch1.json"),
          fetch("./api/branch2.json"),
          fetch("./api/branch3.json"),
        ]);
        const data = await Promise.all(results.map((result) => result.json()));
        logger.info("Got data", data);

        const products = data.map((branchData) => branchData.products).flat();
        logger.info("Products", products);

        const productsById = {};
        for (var product of products) {
          const existingProduct = productsById[product.id];
          if (existingProduct) {
            productsById[product.id] = {
              ...existingProduct,
              sold: existingProduct.sold + product.sold,
            };
          } else {
            productsById[product.id] = product;
          }
        }
        logger.info("Products by id", productsById);

        const sortedProducts = Object.values(productsById);
        sortedProducts.sort((a, b) =>
          a.name > b.name ? 1 : b.name > a.name ? -1 : 0
        );
        logger.info("Sorted products", sortedProducts);

        sortedProducts.forEach((product) => {
          product.revenue = product.sold * product.unitPrice;
        });

        setState({
          uiState: "READY",
          sortedProducts,
        });
      } catch (error) {
        logger.error("Failed to load products", error);
        setState({
          uiState: "ERROR",
          sortedProducts: undefined,
        });
      }
    };

    getData();
  }, []);

  const filteredProducts =
    state.uiState === "READY"
      ? state.sortedProducts.filter((product) =>
          product.name.toLowerCase().includes(filter.toLowerCase())
        )
      : undefined;

  const total =
    state.uiState === "READY"
      ? filteredProducts.reduce((acc, product) => acc + product.revenue, 0)
      : undefined;

  return (
    <section className="product-list">
      <input
        type="text"
        value={filter}
        onChange={(event) => {
          const newFilter = event.target.value;
          logger.info("Setting filter", newFilter);
          setFilter(newFilter);
        }}
      />
      <table>
        <thead>
          <tr>
            <th>Product</th>
            <th>Revenue</th>
          </tr>
        </thead>
        <tbody>
          {state.uiState === "LOADING" && (
            <tr>
              <td colSpan="2" className="loading">
                Loading...
              </td>
            </tr>
          )}
          {state.uiState === "ERROR" && (
            <tr>
              <td colSpan="2" className="error">
                Error loading product list
              </td>
            </tr>
          )}
          {state.uiState === "READY" &&
            filteredProducts.map((product) => (
              <tr key={product.id}>
                <td>{product.name}</td>
                <td>{formatNumber(product.revenue)}</td>
              </tr>
            ))}
        </tbody>
        <tfoot>
          <tr>
            <td>Total</td>
            <td>{state.uiState === "READY" && formatNumber(total)}</td>
          </tr>
        </tfoot>
      </table>
    </section>
  );
};

export default App;
