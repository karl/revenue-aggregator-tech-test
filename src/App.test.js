import React from "react";
import { act } from "react-dom/test-utils";
import Adapter from "enzyme-adapter-react-16";
import { mount, configure } from "enzyme";
import "jest-enzyme";
import branch1 from "../public/api/branch1.json";
import branch2 from "../public/api/branch2.json";
import branch3 from "../public/api/branch3.json";
import App from "./App";

configure({
  adapter: new Adapter(),
});

const responses = {
  "./api/branch1.json": branch1,
  "./api/branch2.json": branch2,
  "./api/branch3.json": branch3,
};

global.fetch = (endpoint) =>
  Promise.resolve({
    json: () => Promise.resolve(responses[endpoint]),
  });

// based on https://blog.pragmatists.com/genuine-guide-to-testing-react-redux-applications-6f3265c11f63
const flushAllPromises = () => new Promise((resolve) => setImmediate(resolve));

export const flushRequestsAndUpdate = async (enzymeWrapper) => {
  await act(async () => {
    await flushAllPromises();
    enzymeWrapper.update();
  });
};

it("renders without crashing", async () => {
  const app = mount(<App />);
  await flushRequestsAndUpdate(app);
});

it("renders loading text initially", async () => {
  const app = mount(<App />);
  expect(app).toIncludeText("Loading...");
  await flushRequestsAndUpdate(app);
});

it("renders a table after data load", async () => {
  const app = mount(<App />);
  expect(app).toIncludeText("Loading...");
  await flushRequestsAndUpdate(app);
  expect(app).not.toIncludeText("Loading...");
});

it("renders rows with id as key", async () => {
  const app = mount(<App />);
  await flushRequestsAndUpdate(app);

  expect(app.find("table tbody tr").at(56).key()).toEqual("098");
  expect(app.find("table tbody tr").at(73).key()).toEqual("047");
});

it("renders table that is sorted ascending", async () => {
  const app = mount(<App />);
  await flushRequestsAndUpdate(app);
  expect(app.find("table")).toMatchSnapshot();
});

it("calculates total revenue of all branches", async () => {
  const app = mount(<App />);
  await flushRequestsAndUpdate(app);
  expect(app.find("tfoot td:last-child").text()).toEqual("2,102,619.44");
});

it("filters the displayed products", async () => {
  const app = mount(<App />);
  await flushRequestsAndUpdate(app);
  const changeEvent = { target: { value: "pear" } };
  app.find("input").simulate("change", changeEvent);
  expect(app.find("tfoot td:last-child").text()).toEqual("60,681.02");
});
