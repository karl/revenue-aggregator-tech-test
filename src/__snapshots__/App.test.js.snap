// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders table that is sorted ascending 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Product
      </th>
      <th>
        Revenue
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      key="082"
    >
      <td>
        Alfalfa Sprouts
      </td>
      <td>
        12,486.74
      </td>
    </tr>
    <tr
      key="027"
    >
      <td>
        Apple
      </td>
      <td>
        14,832.72
      </td>
    </tr>
    <tr
      key="048"
    >
      <td>
        Apricot
      </td>
      <td>
        20,647.26
      </td>
    </tr>
    <tr
      key="068"
    >
      <td>
        Artichoke
      </td>
      <td>
        4,849.95
      </td>
    </tr>
    <tr
      key="110"
    >
      <td>
        Asian Pear
      </td>
      <td>
        11,601.45
      </td>
    </tr>
    <tr
      key="070"
    >
      <td>
        Asparagus
      </td>
      <td>
        20,105.02
      </td>
    </tr>
    <tr
      key="041"
    >
      <td>
        Atemoya
      </td>
      <td>
        11,160.60
      </td>
    </tr>
    <tr
      key="013"
    >
      <td>
        Avocado
      </td>
      <td>
        8,842.50
      </td>
    </tr>
    <tr
      key="054"
    >
      <td>
        Bamboo Shoots
      </td>
      <td>
        11,289.60
      </td>
    </tr>
    <tr
      key="067"
    >
      <td>
        Banana
      </td>
      <td>
        10,872.52
      </td>
    </tr>
    <tr
      key="120"
    >
      <td>
        Bean Sprouts
      </td>
      <td>
        16,626.56
      </td>
    </tr>
    <tr
      key="107"
    >
      <td>
        Beans
      </td>
      <td>
        21,774.33
      </td>
    </tr>
    <tr
      key="111"
    >
      <td>
        Beets
      </td>
      <td>
        25,521.00
      </td>
    </tr>
    <tr
      key="034"
    >
      <td>
        Belgian Endive
      </td>
      <td>
        14,682.80
      </td>
    </tr>
    <tr
      key="129"
    >
      <td>
        Bell Peppers
      </td>
      <td>
        24,284.16
      </td>
    </tr>
    <tr
      key="020"
    >
      <td>
        Bitter Melon
      </td>
      <td>
        17,095.69
      </td>
    </tr>
    <tr
      key="083"
    >
      <td>
        Blackberries
      </td>
      <td>
        28,646.27
      </td>
    </tr>
    <tr
      key="035"
    >
      <td>
        Blueberries
      </td>
      <td>
        12,556.70
      </td>
    </tr>
    <tr
      key="000"
    >
      <td>
        Bok Choy
      </td>
      <td>
        20,254.18
      </td>
    </tr>
    <tr
      key="001"
    >
      <td>
        Boniato
      </td>
      <td>
        9,098.32
      </td>
    </tr>
    <tr
      key="059"
    >
      <td>
        Boysenberries
      </td>
      <td>
        12,843.36
      </td>
    </tr>
    <tr
      key="073"
    >
      <td>
        Broccoflower
      </td>
      <td>
        33,651.20
      </td>
    </tr>
    <tr
      key="062"
    >
      <td>
        Broccoli
      </td>
      <td>
        19,440.30
      </td>
    </tr>
    <tr
      key="060"
    >
      <td>
        Brussels Sprouts
      </td>
      <td>
        3,769.68
      </td>
    </tr>
    <tr
      key="080"
    >
      <td>
        Cabbage
      </td>
      <td>
        11,514.00
      </td>
    </tr>
    <tr
      key="039"
    >
      <td>
        Cactus Pear
      </td>
      <td>
        21,227.68
      </td>
    </tr>
    <tr
      key="105"
    >
      <td>
        Cantaloupe
      </td>
      <td>
        14,853.96
      </td>
    </tr>
    <tr
      key="010"
    >
      <td>
        Carambola
      </td>
      <td>
        9,796.99
      </td>
    </tr>
    <tr
      key="119"
    >
      <td>
        Carrots
      </td>
      <td>
        29,493.04
      </td>
    </tr>
    <tr
      key="003"
    >
      <td>
        Casaba Melon
      </td>
      <td>
        6,220.17
      </td>
    </tr>
    <tr
      key="056"
    >
      <td>
        Cauliflower
      </td>
      <td>
        10,843.26
      </td>
    </tr>
    <tr
      key="008"
    >
      <td>
        Celery
      </td>
      <td>
        9,718.22
      </td>
    </tr>
    <tr
      key="077"
    >
      <td>
        Chayote
      </td>
      <td>
        20,305.76
      </td>
    </tr>
    <tr
      key="066"
    >
      <td>
        Cherimoya
      </td>
      <td>
        15,324.21
      </td>
    </tr>
    <tr
      key="100"
    >
      <td>
        Cherries
      </td>
      <td>
        7,505.74
      </td>
    </tr>
    <tr
      key="131"
    >
      <td>
        Coconuts
      </td>
      <td>
        9,075.53
      </td>
    </tr>
    <tr
      key="030"
    >
      <td>
        Collard Greens
      </td>
      <td>
        16,758.56
      </td>
    </tr>
    <tr
      key="015"
    >
      <td>
        Corn
      </td>
      <td>
        9,437.04
      </td>
    </tr>
    <tr
      key="058"
    >
      <td>
        Cranberries
      </td>
      <td>
        28,912.94
      </td>
    </tr>
    <tr
      key="101"
    >
      <td>
        Cucumber
      </td>
      <td>
        12,106.82
      </td>
    </tr>
    <tr
      key="117"
    >
      <td>
        Dates
      </td>
      <td>
        18,813.60
      </td>
    </tr>
    <tr
      key="019"
    >
      <td>
        Dried Plums
      </td>
      <td>
        10,022.30
      </td>
    </tr>
    <tr
      key="005"
    >
      <td>
        Eggplant
      </td>
      <td>
        10,173.60
      </td>
    </tr>
    <tr
      key="009"
    >
      <td>
        Endive
      </td>
      <td>
        15,671.01
      </td>
    </tr>
    <tr
      key="032"
    >
      <td>
        Escarole
      </td>
      <td>
        8,443.37
      </td>
    </tr>
    <tr
      key="109"
    >
      <td>
        Feijoa
      </td>
      <td>
        12,015.94
      </td>
    </tr>
    <tr
      key="103"
    >
      <td>
        Fennel
      </td>
      <td>
        12,870.70
      </td>
    </tr>
    <tr
      key="074"
    >
      <td>
        Figs
      </td>
      <td>
        14,169.36
      </td>
    </tr>
    <tr
      key="122"
    >
      <td>
        Garlic
      </td>
      <td>
        16,872.96
      </td>
    </tr>
    <tr
      key="099"
    >
      <td>
        Gooseberries
      </td>
      <td>
        2,980.52
      </td>
    </tr>
    <tr
      key="116"
    >
      <td>
        Grapefruit
      </td>
      <td>
        20,697.72
      </td>
    </tr>
    <tr
      key="021"
    >
      <td>
        Grapes
      </td>
      <td>
        26,356.69
      </td>
    </tr>
    <tr
      key="002"
    >
      <td>
        Green Beans
      </td>
      <td>
        11,531.14
      </td>
    </tr>
    <tr
      key="031"
    >
      <td>
        Green Onions
      </td>
      <td>
        15,536.90
      </td>
    </tr>
    <tr
      key="123"
    >
      <td>
        Greens
      </td>
      <td>
        11,121.90
      </td>
    </tr>
    <tr
      key="069"
    >
      <td>
        Guava
      </td>
      <td>
        28,212.45
      </td>
    </tr>
    <tr
      key="098"
    >
      <td>
        Hominy
      </td>
      <td>
        9,083.96
      </td>
    </tr>
    <tr
      key="096"
    >
      <td>
        Honeydew Melon
      </td>
      <td>
        10,559.43
      </td>
    </tr>
    <tr
      key="092"
    >
      <td>
        Horned Melon
      </td>
      <td>
        6,569.64
      </td>
    </tr>
    <tr
      key="004"
    >
      <td>
        Iceberg Lettuce
      </td>
      <td>
        21,069.78
      </td>
    </tr>
    <tr
      key="022"
    >
      <td>
        Jerusalem Artichoke
      </td>
      <td>
        28,203.00
      </td>
    </tr>
    <tr
      key="007"
    >
      <td>
        Jicama
      </td>
      <td>
        13,444.20
      </td>
    </tr>
    <tr
      key="114"
    >
      <td>
        Kale
      </td>
      <td>
        8,933.76
      </td>
    </tr>
    <tr
      key="025"
    >
      <td>
        Kiwifruit
      </td>
      <td>
        33,122.60
      </td>
    </tr>
    <tr
      key="033"
    >
      <td>
        Kohlrabi
      </td>
      <td>
        21,898.89
      </td>
    </tr>
    <tr
      key="036"
    >
      <td>
        Kumquat
      </td>
      <td>
        13,267.92
      </td>
    </tr>
    <tr
      key="028"
    >
      <td>
        Leeks
      </td>
      <td>
        7,269.34
      </td>
    </tr>
    <tr
      key="126"
    >
      <td>
        Lemons
      </td>
      <td>
        17,755.07
      </td>
    </tr>
    <tr
      key="071"
    >
      <td>
        Lettuce
      </td>
      <td>
        10,037.56
      </td>
    </tr>
    <tr
      key="084"
    >
      <td>
        Lima Beans
      </td>
      <td>
        3,670.68
      </td>
    </tr>
    <tr
      key="076"
    >
      <td>
        Limes
      </td>
      <td>
        15,708.00
      </td>
    </tr>
    <tr
      key="081"
    >
      <td>
        Longan
      </td>
      <td>
        18,495.20
      </td>
    </tr>
    <tr
      key="090"
    >
      <td>
        Loquat
      </td>
      <td>
        5,705.68
      </td>
    </tr>
    <tr
      key="047"
    >
      <td>
        Lychee
      </td>
      <td>
        20,981.66
      </td>
    </tr>
    <tr
      key="014"
    >
      <td>
        Madarins
      </td>
      <td>
        4,747.44
      </td>
    </tr>
    <tr
      key="018"
    >
      <td>
        Malanga
      </td>
      <td>
        18,386.10
      </td>
    </tr>
    <tr
      key="087"
    >
      <td>
        Mandarin Oranges
      </td>
      <td>
        12,380.40
      </td>
    </tr>
    <tr
      key="061"
    >
      <td>
        Mangos
      </td>
      <td>
        10,977.28
      </td>
    </tr>
    <tr
      key="042"
    >
      <td>
        Mulberries
      </td>
      <td>
        9,189.96
      </td>
    </tr>
    <tr
      key="075"
    >
      <td>
        Mushrooms
      </td>
      <td>
        12,557.84
      </td>
    </tr>
    <tr
      key="104"
    >
      <td>
        Napa
      </td>
      <td>
        15,087.56
      </td>
    </tr>
    <tr
      key="125"
    >
      <td>
        Nectarines
      </td>
      <td>
        31,706.88
      </td>
    </tr>
    <tr
      key="086"
    >
      <td>
        Okra
      </td>
      <td>
        20,784.21
      </td>
    </tr>
    <tr
      key="029"
    >
      <td>
        Onion
      </td>
      <td>
        15,888.04
      </td>
    </tr>
    <tr
      key="043"
    >
      <td>
        Oranges
      </td>
      <td>
        10,575.18
      </td>
    </tr>
    <tr
      key="051"
    >
      <td>
        Papayas
      </td>
      <td>
        19,369.74
      </td>
    </tr>
    <tr
      key="016"
    >
      <td>
        Parsnip
      </td>
      <td>
        17,594.65
      </td>
    </tr>
    <tr
      key="118"
    >
      <td>
        Passion Fruit
      </td>
      <td>
        21,589.92
      </td>
    </tr>
    <tr
      key="023"
    >
      <td>
        Peaches
      </td>
      <td>
        18,759.18
      </td>
    </tr>
    <tr
      key="046"
    >
      <td>
        Pears
      </td>
      <td>
        10,329.30
      </td>
    </tr>
    <tr
      key="050"
    >
      <td>
        Peas
      </td>
      <td>
        17,965.60
      </td>
    </tr>
    <tr
      key="093"
    >
      <td>
        Peppers
      </td>
      <td>
        8,838.01
      </td>
    </tr>
    <tr
      key="024"
    >
      <td>
        Persimmons
      </td>
      <td>
        24,514.18
      </td>
    </tr>
    <tr
      key="115"
    >
      <td>
        Pineapple
      </td>
      <td>
        23,462.40
      </td>
    </tr>
    <tr
      key="053"
    >
      <td>
        Plantains
      </td>
      <td>
        9,978.90
      </td>
    </tr>
    <tr
      key="017"
    >
      <td>
        Plums
      </td>
      <td>
        13,314.60
      </td>
    </tr>
    <tr
      key="045"
    >
      <td>
        Pomegranate
      </td>
      <td>
        13,551.09
      </td>
    </tr>
    <tr
      key="127"
    >
      <td>
        Potatoes
      </td>
      <td>
        10,812.54
      </td>
    </tr>
    <tr
      key="044"
    >
      <td>
        Prickly Pear
      </td>
      <td>
        17,522.59
      </td>
    </tr>
    <tr
      key="038"
    >
      <td>
        Prunes
      </td>
      <td>
        13,644.54
      </td>
    </tr>
    <tr
      key="130"
    >
      <td>
        Pummelo
      </td>
      <td>
        20,429.40
      </td>
    </tr>
    <tr
      key="097"
    >
      <td>
        Pumpkin
      </td>
      <td>
        10,547.88
      </td>
    </tr>
    <tr
      key="102"
    >
      <td>
        Quince
      </td>
      <td>
        15,250.34
      </td>
    </tr>
    <tr
      key="112"
    >
      <td>
        Radicchio
      </td>
      <td>
        20,453.27
      </td>
    </tr>
    <tr
      key="106"
    >
      <td>
        Radishes
      </td>
      <td>
        30,051.48
      </td>
    </tr>
    <tr
      key="078"
    >
      <td>
        Raisins
      </td>
      <td>
        7,411.05
      </td>
    </tr>
    <tr
      key="011"
    >
      <td>
        Raspberries
      </td>
      <td>
        4,318.82
      </td>
    </tr>
    <tr
      key="064"
    >
      <td>
        Red Cabbage
      </td>
      <td>
        12,005.60
      </td>
    </tr>
    <tr
      key="079"
    >
      <td>
        Rhubarb
      </td>
      <td>
        14,905.80
      </td>
    </tr>
    <tr
      key="085"
    >
      <td>
        Romaine Lettuce
      </td>
      <td>
        32,786.28
      </td>
    </tr>
    <tr
      key="095"
    >
      <td>
        Rutabaga
      </td>
      <td>
        18,551.12
      </td>
    </tr>
    <tr
      key="094"
    >
      <td>
        Shallots
      </td>
      <td>
        6,517.17
      </td>
    </tr>
    <tr
      key="108"
    >
      <td>
        Snow Peas
      </td>
      <td>
        11,989.19
      </td>
    </tr>
    <tr
      key="012"
    >
      <td>
        Spinach
      </td>
      <td>
        19,822.32
      </td>
    </tr>
    <tr
      key="040"
    >
      <td>
        Sprouts
      </td>
      <td>
        27,061.80
      </td>
    </tr>
    <tr
      key="088"
    >
      <td>
        Squash
      </td>
      <td>
        15,080.00
      </td>
    </tr>
    <tr
      key="063"
    >
      <td>
        Strawberries
      </td>
      <td>
        11,681.46
      </td>
    </tr>
    <tr
      key="124"
    >
      <td>
        String Beans
      </td>
      <td>
        22,731.62
      </td>
    </tr>
    <tr
      key="128"
    >
      <td>
        Sweet Potato
      </td>
      <td>
        31,680.00
      </td>
    </tr>
    <tr
      key="072"
    >
      <td>
        Tangelo
      </td>
      <td>
        19,831.35
      </td>
    </tr>
    <tr
      key="052"
    >
      <td>
        Tangerines
      </td>
      <td>
        29,450.24
      </td>
    </tr>
    <tr
      key="026"
    >
      <td>
        Tomatillo
      </td>
      <td>
        9,560.46
      </td>
    </tr>
    <tr
      key="113"
    >
      <td>
        Tomato
      </td>
      <td>
        14,517.90
      </td>
    </tr>
    <tr
      key="121"
    >
      <td>
        Turnip
      </td>
      <td>
        15,564.12
      </td>
    </tr>
    <tr
      key="049"
    >
      <td>
        Ugli Fruit
      </td>
      <td>
        10,838.10
      </td>
    </tr>
    <tr
      key="057"
    >
      <td>
        Water Chestnuts
      </td>
      <td>
        7,857.66
      </td>
    </tr>
    <tr
      key="006"
    >
      <td>
        Watercress
      </td>
      <td>
        13,642.14
      </td>
    </tr>
    <tr
      key="091"
    >
      <td>
        Watermelon
      </td>
      <td>
        19,219.20
      </td>
    </tr>
    <tr
      key="132"
    >
      <td>
        Waxed Beans
      </td>
      <td>
        9,119.88
      </td>
    </tr>
    <tr
      key="089"
    >
      <td>
        Yams
      </td>
      <td>
        13,719.66
      </td>
    </tr>
    <tr
      key="065"
    >
      <td>
        Yellow Squash
      </td>
      <td>
        23,634.52
      </td>
    </tr>
    <tr
      key="037"
    >
      <td>
        Yuca/Cassava
      </td>
      <td>
        44,490.84
      </td>
    </tr>
    <tr
      key="055"
    >
      <td>
        Zucchini Squash
      </td>
      <td>
        13,077.28
      </td>
    </tr>
  </tbody>
  <tfoot>
    <tr>
      <td>
        Total
      </td>
      <td>
        2,102,619.44
      </td>
    </tr>
  </tfoot>
</table>
`;
